crates/ruff_server/src/server.rs

```diff
@@ -0,0 +1,152 @@
+
+use self::schedule::main_thread;
+use crate::session::Session;
+
+mod api;
+mod client;
+mod schedule;
+
+pub(crate) type Result<T> = std::result::Result<T, api::Error>;
+
+pub struct Server {
+    conn: lsp::Connection,
+    threads: lsp::IoThreads,
+    session: Session,
+}
+
+impl Server {
+    pub fn new() -> crate::Result<Self> {
+        let (conn, threads) = lsp::Connection::stdio();
+
+        let (id, params) = conn.initialize_start()?;
+
+        let init_params: types::InitializeParams = serde_json::from_value(params)?;
+
+        let client_capabilities = init_params.capabilities;
+        let server_capabilities = Self::server_capabilities(&client_capabilities);
+
+        let workspaces = init_params
+            .workspace_folders
+            .map(|f| f.into_iter().map(|w| w.uri).collect())
+            .or_else(|| init_params.root_uri.map(|u| vec![u])).ok_or_else(|| anyhow!("No workspace or root URI was given in the LSP initialization parameters. The server cannot start."))?;
+
+        let initialize_data = serde_json::json!({
+            "capabilities": server_capabilities,
+            "serverInfo": {
+                "name": crate::SERVER_NAME,
+                "version": crate::version()
+            }
+        });
+
+        conn.initialize_finish(id, initialize_data)?;
+
+        Ok(Self {
+            conn,
+            threads,
+            session: Session::new(&server_capabilities, &workspaces)?,
+        })
+    }
+
+    pub fn run(self) -> crate::Result<()> {
+        let result = main_thread(move || Self::main_loop(&self.conn, self.session))?.join();
+        self.threads.join()?;
+        result
+    }
+
+    fn main_loop(connection: &Connection, session: Session) -> crate::Result<()> {
+        // TODO(jane): Make thread count configurable
+        let mut scheduler = schedule::Scheduler::new(session, 4, &connection.sender);
+        for msg in &connection.receiver {
+            scheduler.process_events();
+            let task = match msg {
+                lsp::Message::Request(req) => {
+                    if connection.handle_shutdown(&req)? {
+                        return Ok(());
+                    }
+                    api::request(req)
+                }
+                lsp::Message::Notification(notification) => api::notification(notification),
+                lsp::Message::Response(response) => {
+                    tracing::error!(
+                        "Expected request or notification, got response instead: {response:?}"
+                    );
+                    continue;
+                }
+            };
+            scheduler.dispatch(task);
+        }
+        Ok(())
+    }
+
+    fn server_capabilities(client_capabilities: &ClientCapabilities) -> types::ServerCapabilities {
+        let position_encoding = client_capabilities
+            .general
+            .as_ref()
+            .and_then(|g| g.position_encodings.as_ref())
+            .and_then(|e| e.first())
```

Comment:
Nit: I would appreciate longer names than `g` and `e`. It improves readability when you don't have type information available, e.g. on GitHub.